### PR TITLE
[FIX,AUTOSCHEDULER] Fix auto_scheduler to run with multiprocessing's spawn start method

### PR DESCRIPTION
**Pull request by tkonolige**
This PR fixes autotvm and the autoscheduler when using multiprocessing's spawn start method. I had to remove all nested function declarations, property serialize auto_scheduler's `SearchTask`, and propagate registries to subprocesses. Unfortunately, the registry propagation does not work correctly when running under pytest. I've disable those tests for now (`tests/python/unittest/test_runtime_rpc.py`), but they work if run directly. Maybe someone else can take a look at this.

See #6650 

@merrymercy

### CODE REVIEW

**Comment by merrymercy** [comment]
Note that "spawn" is slower than "fork" (https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods).
We should not change all default settings to "spawn".
We should prefer "fork" as the default setting when it is possible.

**Comment by tkonolige** [comment]
The problem with using fork is that it is unsafe on macOS or when using the CUDA api. The pytorch developers recommend not using fork with CUDA: https://pytorch.org/docs/master/notes/multiprocessing.html#cuda-in-multiprocessing. It looks like we have to choose something that works over something that is fast.

I also don't think it really matters that spawning is slower than forking. We are not spawning that many processes. And if the overhead from spawning is big enough to cause a significant slowdown, then we shouldn't have been using processes/threads in the first place.

**Comment by merrymercy** [comment]
Autotvm uses a lot of python multiprocessing and I expect it will be much slower when using spawn.  AutoTVM uses multiprocessing for feature extraction. So it needs to launch about 50,000 processes every measurement batch.

Ansor does not rely on multiprocessing too much. So the situation for Ansor is better.
However, I'd like to see benchmark numbers about fork vs. spawn.
i.e. Run a search with `n-trials=64` with (Ansor, AutoTVM) x (fork, spawn) X (CPU, GPU) (Optional)

**Comment by tkonolige** [comment]
@merrymercy spawn is definitely slower. With autotvm, spawn is about 50% slower. I figured out where the CUDA drivers were being called in forked threads and fixed it. Now fork should be used on linux.

**Comment by tkonolige** [comment]
@merrymercy @tqchen A review on this would be great.

**Comment by tqchen** [comment]
cc @jcf94 @comaniac @yzhliu please also help to take a look

File: python/tvm/autotvm/tuner/xgboost_cost_model.py [hunk 1]
```diff
@@ -321,10 +316,11 @@ def _get_feature(self, indexes):
 
         indexes = np.array(indexes)
         need_extract = [x for x in indexes if x not in fea_cache]
+        args = [(self.space.get(x), self.target, self.task) for x in need_extract]
```
Comment (merrymercy) [inline python/tvm/autotvm/tuner/xgboost_cost_model.py:319]: Doing this still needs serializing a lot of things. Did you test the performance before vs. after?

File: src/auto_scheduler/measure_record.cc [hunk 2]
```diff
@@ -107,18 +107,68 @@ struct Handler<::tvm::auto_scheduler::StateNode> {
...
+};
+
 template <>
 struct Handler<::tvm::auto_scheduler::SearchTaskNode> {
   inline static void Write(dmlc::JSONWriter* writer,
                            const ::tvm::auto_scheduler::SearchTaskNode& data) {
     writer->BeginArray(false);
     writer->WriteArrayItem(std::string(data.workload_key));
     writer->WriteArrayItem(data.target->str());
+    writer->WriteArrayItem(*data.hardware_params.get());
```
Comment (merrymercy) [inline src/auto_scheduler/measure_record.cc:165]: Since you changed the format, please update the version number to `v0.3`
https://github.com/apache/incubator-tvm/blob/c6f18250e176a0f107481d489651f6abdfe00976/src/auto_scheduler/measure_record.cc#L219

File: python/tvm/auto_scheduler/utils.py [hunk 3]
```diff
@@ -169,17 +143,18 @@ def kill_child_processes(parent_pid, sig=signal.SIGTERM):
             return
 
 
+def _func_wrapper(que, func, args, kwargs):
```
Comment (merrymercy) [inline python/tvm/auto_scheduler/utils.py:146]: document it?

File: python/tvm/auto_scheduler/measure.py [hunk 4]
```diff
@@ -87,6 +120,24 @@ def __init__(self, task, state):
         state = state if isinstance(state, StateObject) else state.state_object
         self.__init_handle_by_constructor__(_ffi_api.MeasureInput, task, state)
 
+    def serialize(self):
```
Comment (merrymercy) [inline python/tvm/auto_scheduler/measure.py:123]: Can we use `__getstate__`?

File: python/tvm/auto_scheduler/measure.py [hunk 5]
```diff
@@ -590,14 +634,20 @@ def local_builder_build(inputs, timeout, n_parallel, build_func="default", verbo
...
-    # We use fork and a global variable to copy arguments between processes.
-    # This can avoid expensive serialization of TVM IR when using multiprocessing.Pool
-    global GLOBAL_BUILD_ARGUMENTS
-
-    GLOBAL_BUILD_ARGUMENTS = (inputs, build_func, timeout, verbose)
-
-    pool = NoDaemonPool(n_parallel)
-    tuple_res = pool.map(local_build_worker, range(len(inputs)))
+    # This pool is not doing computationally intensive work, so we can use threads
+    pool = multiprocessing.pool.ThreadPool(n_parallel)
```
Comment (merrymercy) [inline python/tvm/auto_scheduler/measure.py:638]: Do we still need serialization with `ThreadPool`?
I guess all these arguments will be shared in memory and will be passed as references.

**Review by merrymercy — CHANGES_REQUESTED** [review]

**Comment by merrymercy** [comment]
Thanks for the refactoring! I would like to see benchmark results for both autotvm and auto-scheduler, so we don't get performance regression.

File: python/tvm/autotvm/task/task.py [hunk 6]
```diff
@@ -173,6 +172,8 @@ def __getstate__(self):
         # some unpickable local task functions.
         # So we only pickle the name of the function
         # and restore the function by name when unpickling it.
+        import cloudpickle  # pylint: disable=import-outside-toplevel
```
[the remark this answers is not in the document]
Comment (tqchen) [reply python/tvm/autotvm/task/task.py:175]: in this case I asked @tkonolige to import here for now since cloudpickle is an optional dep introduced by autotuning atm,

File: python/tvm/auto_scheduler/workload_registry.py [hunk 7]
```diff
@@ -175,6 +175,41 @@ def workload_key_to_tensors(workload_key):
...
+    workload : callable
+        The registered workload function.
+    """
+    name = workload_name(task.workload_key)
+    lookup = WORKLOAD_FUNC_REGISTRY[name]
+    assert callable(lookup)
+    return lookup
+
+
+def workload_name(workload_key):
```
Comment (merrymercy) [inline python/tvm/auto_scheduler/workload_registry.py:197]:
```suggestion
def workload_func_name(workload_key):
```

**Review by merrymercy — CHANGES_REQUESTED** [review]

**Comment by merrymercy** [comment]
Another concern is that I introduced a new kind of workload in this PR (#6710) for relay integration.
When using auto-scheduler in Relay, we don't serialize a ComputeDAG by function + arguments.
We save a copy of the TE expressions generated by relay as the workload.
So we have to support serializing TE expressions after this PR. I am not sure if it works because there are pointers in TE expressions.

File: python/tvm/auto_scheduler/measure.py (hunk 4 above)
Comment (tkonolige) [reply python/tvm/auto_scheduler/measure.py:123]: There was some weird initialization bug with `__getstate__` that I could not figure out. I'll add a comment about this.

File: python/tvm/auto_scheduler/measure.py [hunk 8]
```diff
@@ -950,26 +1017,30 @@ def rpc_runner_run(
...
-        min_repeat_ms,
-        cooldown_interval,
-        enable_cpu_cache_flush,
-        verbose,
-    )
-
     assert len(inputs) == len(build_results), "Measure input size should be equal to build results"
-    pool = NoDaemonPool(n_parallel)
-    tuple_res = pool.map(rpc_run_worker, range(len(build_results)))
+    # This pool is not doing computationally intensive work, so we can use threads
```
[the remark this answers is not in the document]
Comment (tkonolige) [reply python/tvm/auto_scheduler/measure.py:1021]: The pool is doing basically no work. Each thread in the pool spawns a process and then waits till it times out.

**Comment by tkonolige** [comment]
Here are benchmark results on linux using forking. This is `dense` with 64 x 64 matrices. ntrials=16, 5 bencmarking iterations.

| Branch | Auto-what | Target | Mean Time | STD |
| ------ | --------- | ------ | --------- | --- |
| main | autotvm        | llvm  | 20.429478 | 1.914881 |
| PR | autotvm       | llvm | 26.400752 | 9.013005 |
| main | autotvm        | cuda  |  7.010600 | 0.080864 |
| PR | autotvm       | cuda |  7.001349 | 0.017357 |
| main | autoscheduler  | llvm  | 16.711073 | 2.572820 |
| PR | autoscheduler | llvm | 14.752735 | 0.757895 |
| main | autoscheduler  | cuda  | 32.981713 | 0.072185 |
| PR | autoscheduler | cuda | 32.283536 | 0.054209 |

Looks pretty similar except for high standard deviation in runtime. Not sure what's causing this. I still think we should merge this as it fixes users not being able to run auto scheduling on Macs.

**Review by merrymercy — DISMISSED** [review]

**Comment by merrymercy** [comment]
The auto-scheduler part looks good to me.
But I am also not sure about the autotvm part.
If you don't mind, I suggest deleting the autotvm part in this PR.

**Review by jcf94 — COMMENTED** [review]
Thanks for your contribution! I don't have more comments on this.

**Comment by tkonolige** [comment]
@merrymercy The autotvm changes are required for things to work.

**Comment by tkonolige** [comment]
Here are the autotvm times you requested. 1024x1024 with 100 trials.
```
This PR
autotvm       llvm 297.038928
autotvm       cuda 15.138037

main
autotvm       llvm 297.375458
autotvm       cuda 14.163432
```

Results are very similar.

**Comment by merrymercy** [comment]
I tested your autotvm implementation and found it is significantly slower.

To do the test, you can run this tutorial (https://github.com/apache/incubator-tvm/blob/main/tutorials/autotvm/tune_conv2d_cuda.py) and replace the n_trial=20 with n_trial=100

The number we care about is the time spent on simulated annealing which actually uses the xgboost_cost_model.py modified by you.

This is the screenshot before your PR. You can see elapsed: 63.40 in the last line, which means it takes 63.40 ms to do simulated annealing.
image

However, with your PR, the number becomes 194.67

Screen Shot 2020-10-27 at 12 47 15 PM

I don’t understand why the autotvm part is required for that PR, because autotvm and auto-scheduler are totally independent.  Can you delete autotvm files in that PR?

The test machine is a 24-core Intel E5

**Comment by tkonolige** [comment]
@merrymercy The goal of this pr is to make autotuning and auto scheduling work on macOS and windows. Currently autotvm does not work on macOS, so that is why I have been touching it.

The tutorial you are using is broken. All code that needs to be run should be within an `if __name__ == "__main__":` statement. With this statement, and changing n_trials to 100, I get these results:

On main
```
SA iter: 50	last_update: 49	max-0: 5.05	max-1: 5.32	temp: 0.90	elapsed: 22.72
SA iter: 100	last_update: 98	max-0: 5.19	max-1: 5.43	temp: 0.80	elapsed: 44.25
SA iter: 150	last_update: 147	max-0: 5.19	max-1: 5.43	temp: 0.70	elapsed: 65.84
SA iter: 200	last_update: 198	max-0: 5.29	max-1: 5.43	temp: 0.60	elapsed: 87.38
SA iter: 250	last_update: 248	max-0: 5.29	max-1: 5.43	temp: 0.50	elapsed: 108.47
SA iter: 300	last_update: 299	max-0: 5.31	max-1: 5.43	temp: 0.40	elapsed: 129.17
SA iter: 350	last_update: 331	max-0: 5.32	max-1: 5.43	temp: 0.30	elapsed: 149.26
SA iter: 400	last_update: 396	max-0: 5.32	max-1: 5.43	temp: 0.20	elapsed: 168.95
SA iter: 450	last_update: 448	max-0: 5.32	max-1: 5.43	temp: 0.10	elapsed: 187.60
SA iter: 500	last_update: 499	max-0: 5.32	max-1: 5.43	temp: 0.00	elapsed: 204.33
SA iter: 500	last_update: 499	elapsed: 204.33
```

On this PR
```
SA iter: 50	last_update: 49	max-0: 6.39	max-1: 6.87	temp: 0.90	elapsed: 23.53
SA iter: 100	last_update: 97	max-0: 6.56	max-1: 7.05	temp: 0.80	elapsed: 46.57
SA iter: 150	last_update: 149	max-0: 6.68	max-1: 7.05	temp: 0.70	elapsed: 69.27
SA iter: 200	last_update: 192	max-0: 6.69	max-1: 7.05	temp: 0.60	elapsed: 91.90
SA iter: 250	last_update: 238	max-0: 6.71	max-1: 7.05	temp: 0.50	elapsed: 114.53
SA iter: 300	last_update: 294	max-0: 6.77	max-1: 7.05	temp: 0.40	elapsed: 136.67
SA iter: 350	last_update: 348	max-0: 6.77	max-1: 7.05	temp: 0.30	elapsed: 158.47
SA iter: 400	last_update: 398	max-0: 6.87	max-1: 7.05	temp: 0.20	elapsed: 179.36
SA iter: 450	last_update: 447	max-0: 6.87	max-1: 7.05	temp: 0.10	elapsed: 198.81
SA iter: 500	last_update: 497	max-0: 6.87	max-1: 7.05	temp: 0.00	elapsed: 216.06
SA iter: 500	last_update: 497	elapsed: 216.06
```

Which has the runtime pretty close (I'm not sure what the variance on times is here).

**Comment by merrymercy** [comment]
What's your testing environment?

**Comment by tkonolige** [comment]
If you don't have the `if __name__ == "__main__":`, then the the whole body of the module may be run on every sub process. Could you test on your machine with this change?

My environment is archlinux with a Intel(R) Core(TM) i5-3570K CPU @ 3.40GHz and 1080. Python 3.8.6.

**Comment by merrymercy** [comment]
Which PR introduced the changes that broke the tutorials? I don't find related thing in your PR.

**Comment by tkonolige** [comment]
They are broken with respect to the multiprocessing changes I have introduced in this PR. Technically speaking they were never correct as the script to be executed should be wrapped in `__name__ == "__main__"`.

**Comment by merrymercy** [comment]
Sorry, could you point me to the related lines? I only see you rename some variables in the `python/tvm/autotvm/measure/local_executor.py`

**Comment by tkonolige** [comment]
I took off the `__name__ == "__main__"` lines and it performs the same. `__name__ == "__main__"` is still required for non-linux machines (i.e. the tutorials are buggy on non linux machines), but it isn't causing the issue here. The only difference I can see then is that your machine has so many threads. Can you reproduce your results on a different machine?

**Comment by tkonolige** [comment]
Here is the tests I've run on a 64-core machine:

```
This PR
XGB iter:   0	tr-a-recall@64: 0.598716	tr-map: 0.142857
XGB iter:  25	tr-a-recall@64: 0.667324	tr-map: 0.500000
XGB stopped. Best iteration: [13] tr-a-recall@64:0.67690	tr-map:0.50000
XGB train: 1.24	obs: 64	error: 50	n_cache: 64
SA iter: 50	last_update: 49	max-0: 4.42	max-1: 4.71	temp: 0.90	elapsed: 16.14
SA iter: 100	last_update: 97	max-0: 4.55	max-1: 4.71	temp: 0.80	elapsed: 30.59
SA iter: 150	last_update: 145	max-0: 4.56	max-1: 4.74	temp: 0.70	elapsed: 45.74
SA iter: 200	last_update: 197	max-0: 4.58	max-1: 4.83	temp: 0.60	elapsed: 60.61
SA iter: 250	last_update: 249	max-0: 4.67	max-1: 4.84	temp: 0.50	elapsed: 76.09
SA iter: 300	last_update: 299	max-0: 4.69	max-1: 4.84	temp: 0.40	elapsed: 90.07
SA iter: 350	last_update: 349	max-0: 4.71	max-1: 4.84	temp: 0.30	elapsed: 104.27
SA iter: 400	last_update: 377	max-0: 4.71	max-1: 4.84	temp: 0.20	elapsed: 118.12
SA iter: 450	last_update: 448	max-0: 4.72	max-1: 4.84	temp: 0.10	elapsed: 130.86
SA iter: 500	last_update: 498	max-0: 4.76	max-1: 4.84	temp: 0.00	elapsed: 143.39
SA iter: 500	last_update: 498	elapsed: 143.39

main
XGB iter:   0	tr-a-recall@64: 0.641594	tr-map: 0.500000
XGB iter:  25	tr-a-recall@64: 0.716034	tr-map: 0.500000
XGB iter:  50	tr-a-recall@64: 0.728245	tr-map: 1.000000
XGB stopped. Best iteration: [38] tr-a-recall@64:0.73020	tr-map:0.50000
XGB train: 1.10	obs: 64	error: 46	n_cache: 64
SA iter: 50	last_update: 49	max-0: 6.21	max-1: 6.94	temp: 0.90	elapsed: 3.40
SA iter: 100	last_update: 99	max-0: 6.50	max-1: 7.16	temp: 0.80	elapsed: 6.68
SA iter: 150	last_update: 149	max-0: 6.74	max-1: 7.38	temp: 0.70	elapsed: 10.24
SA iter: 200	last_update: 199	max-0: 6.86	max-1: 7.39	temp: 0.60	elapsed: 13.91
SA iter: 250	last_update: 249	max-0: 6.97	max-1: 7.39	temp: 0.50	elapsed: 17.68
SA iter: 300	last_update: 294	max-0: 7.07	max-1: 7.56	temp: 0.40	elapsed: 21.29
SA iter: 350	last_update: 348	max-0: 7.11	max-1: 7.56	temp: 0.30	elapsed: 24.52
SA iter: 400	last_update: 399	max-0: 7.15	max-1: 7.56	temp: 0.20	elapsed: 27.68
SA iter: 450	last_update: 449	max-0: 7.16	max-1: 7.56	temp: 0.10	elapsed: 30.63
SA iter: 500	last_update: 499	max-0: 7.17	max-1: 7.56	temp: 0.00	elapsed: 33.48
SA iter: 500	last_update: 499	elapsed: 33.48
```
Seems like the issue is hit when there are a lot of cores.

I've pulled out the autotvm parts and will send another pr when I figure out those.

**Comment by merrymercy** [comment]
How about also undoing the tutorials? I guess their format will be wrong because you add indentation to text blocks.
For the autotvm problem, I am okay to accept the following solution:
1. provide two versions of the code.
2. pick the version according to OS.

**Comment by tqchen** [comment]
I agree that the sphinx gallery will run all tutorials together, and it would still be useful to keep the tutorials in the form of ipynb style(no main)

**Comment by tkonolige** [comment]
Maybe we should at least leave a comment that they are broken when using the spawn method?

**Review by merrymercy — APPROVED** [review]
